chore(docs): Algolia search guide from class to functional components (#14144)
gatsbyjs/gatsby: docs/docs/adding-search-with-algolia.md

title: Adding search with Algolia
---

Once you've added some content to your site, you'll want to make it easy for your visitors to find what they're looking for. This guide will run you through the process of setting up a custom search experience powered by [Algolia](https://www.algolia.com) on any Gatsby site.

Two things before we begin:

1. Beyond this guide, you may also want to checkout Algolia's extensive [docs on how to get started in React](https://www.algolia.com/doc/guides/building-search-ui/getting-started/react).
2. If you're looking to add search to a documentation site, you can let Algolia handle most of the steps outlined below by using their [Docsearch](https://community.algolia.com/docsearch) functionality. For other types of sites and more fine-grained control over exactly what data should be indexed, read on.

## Configuring the Algolia plugin

> First, you'll need to add [`gatsby-plugin-algolia`](https://github.com/algolia/gatsby-plugin-algolia) and [`react-instantsearch-dom`](https://github.com/algolia/react-instantsearch) to your project. `react-instantsearch` is Algolia's library containing off-the-shelf React components which we can import to save ourselves a lot of work. If you're not using it already, also install [`dotenv`](https://github.com/motdotla/dotenv) while you're at it. We're going to need it to specify your Algolia app ID and both the search and admin API keys without commiting them to version control.

Reviewer: ```suggestion
First, you'll need to add [`gatsby-plugin-algolia`](https://github.com/algolia/gatsby-plugin-algolia) and [`react-instantsearch-dom`](https://github.com/algolia/react-instantsearch) to your project. `react-instantsearch` is Algolia's library containing off-the-shelf React components which you can import to save ourselves a lot of work. You'll also be using `dotenv` which gets shipped with Gatsby by default. You're going to need it to specify your Algolia app ID and both the search and admin API keys without commiting them to version control.
```

dotenv already comes with Gatsby. Also: https://www.gatsbyjs.org/contributing/gatsby-style-guide/#use-you-as-the-pronoun

Reviewer 2: @LekoArts it's a bit non-ideal to rely on downstream dependencies to exist (and be the expected version), so I'd keep the installation of dotenv.

Reviewer: mh, interesting. We do the same (relying on that fact) everywhere else I think, e.g. in the env variables documentation.

Author: I'm happy to go either way but I agree with @DSchau that it's often good to keep dependencies explicit. Of course, consistency is nice as well.

> ```sh

Reviewer: Please keep `shell`

```suggestion
```shell
```
https://www.gatsbyjs.org/contributing/gatsby-style-guide/#format-code-blocks-inline-code-and-images

> yarn add gatsby-plugin-algolia react-instantsearch-dom algoliasearch dotenv

Reviewer: ```suggestion
npm install --save gatsby-plugin-algolia react-instantsearch-dom algoliasearch
```

```

> We will be using `styled-components` to design the search UI in this guide but you can use whichever CSS solution you prefer. If you'd like to start using `styled-components` as well, you also need to install

Reviewer: ```suggestion
You will be using `styled-components` to design the search UI in this guide but you can use whichever CSS solution you prefer. If you'd like to start using `styled-components` as well, you also need to install
```

> ```sh

Reviewer: ```suggestion
```shell
```

> yarn add react styled-components gatsby-plugin-styled-components

Reviewer: ```suggestion
npm install --save styled-components gatsby-plugin-styled-components
```

```

> Next, add `gatsby-plugin-algolia` and `gatsby-plugin-styled-components` if you installed it to your `gatsby-config.js`.

Reviewer: ```suggestion
Next, add `gatsby-plugin-algolia` and `gatsby-plugin-styled-components` to your `gatsby-config.js`.
```

        chunkSize: 10000, // default: 1000
      },
    },
    `gatsby-plugin-styled-components`,
  ],
}
```

Notice that we're loading `queries` from a file at `./src/utils/algolia.js` (you can of course put it wherever you like) and our Algolia ID and API key from `.env` so let's add those files.

> ```sh:title=.env

Reviewer: ```suggestion
```text:title=.env
```

Author: I usually get better syntax highlighting (e.g. of comments) in `.env` files with `sh` but I assume `text` is convention?

        frontmatter {
          title
          slug

>           date(formatString: "MMM D, YYYY")

Reviewer: Why this change?

Author: Looks better in my opinion. Just a small oversight during the original writing of this guide but we can also keep it as is.

Reviewer: Okay 👍

## Adding a search interface to your site

Next, let's build a user-facing search interface for your site. It needs a way for the user to enter a search string, send that string to Algolia, receive matching results (_hits_ in Algolia speak) from your indices and finally display those to the user. Let's dive right in.

We're going to assemble everything we need into a React `Search` component that we call from anywhere on our site where we want the user to be able to search. Even though design varies strongly from site to site, I'll also go through the styles implemented with [`styled-components`](https://styled-components.com) in this guide since working out the CSS transitions to have the search field slide out as the user clicks on it and the results pane to appear once Algolia returns matches took some time.

The `Search` components is made up of the following files:

- **`index.js`**: the main component
- **`input.js`**: the text input field
- **`hitComps.js`**: the components that will render matching posts/pages
- **`styles.js`**: the styled components

There's quite a lot happening in these files so let's break them down one by one and piece by piece.

### `index.js`

> ```jsx:title=src/components/search/index.js

Reviewer: ```suggestion
```js:title=src/components/search/index.js
```

))
```

The `Input` component is where the user enters the search string. It is quite short since the grunt work is done by Algolia's [`connectSearchBox`](https://community.algolia.com/react-instantsearch/connectors/connectSearchBox.html) function.

Now let's look at the styled components `SearchIcon`, `Form`, `Input` as well as the ones imported in `index.js`.

## `styled.js`

> ```jsx:title=src/components/search/styles.js

Reviewer: ```suggestion
```js:title=src/components/search/styles.js
```
